modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+
+  The following operations may invalidate references to elements contained in
+  a List:
+
+      - insert
+      - remove
+      - sort
+      - pop
+      - clear
+
+  Additionally, all references to List elements are invalidated when the List
+  is deinitialized.
+
+  List objects are not parallel safe by default, but can be made parallel safe
+  by setting the param formal `parSafe` to true in any List constructor. When
+  constructed from another List, a List object will inherit the parallel
+  safety mode of its originating List.
+
+  Inserts and removals into a List are O(n) worst case and should be performed
+  with care. Access into a List object is currently O(log n), but is expected
+  to reduce to O(1) in the future.
+*/
+module Lists {
+
+  pragma "no doc"
+  config const _initialCapacity = 8;
+
+  pragma "no doc"
+  config const _initialBlockCapacity = 16;
+
+  pragma "no doc"
+  config param _sanityChecks = true;
+
+  //
+  // Some asserts are useful while developing, but can be turned off when the
+  // implementation is correct.
+  //
+  private inline proc _sanity(expr: bool) {
+    if _sanityChecks then
+      assert(expr);
+  }
+
+  //
+  // Basically just a wrapper around _ddata at this point, with one based
+  // indexing to avoid logical errors.
+  //
+  pragma "no doc"
+  class ListBlock {
+
+    type _etype;
+    var _data: _ddata(_etype);
+    const capacity: int;
+
+    proc init(type _etype, capacity: int) {
+      _sanity(capacity > 0);
+      this._etype = _etype;
+      //
+      // Use `c_calloc` to avoid calling initializers. TODO: Is use of calloc
+      // safe in a multi-locale context? Michael said that the cast to
+      // _ddata will convert the c_ptr to a wide pointer.
+      //
+      this._data = c_calloc(_etype, capacity):_ddata(_etype);
+      this.capacity = capacity;
+    }
+
+    proc deinit() {
+      c_free(_data:c_void_ptr);
+    }
+
+    inline proc this(i: int) ref {
+      _sanity(i >= 1 && i <= capacity);
+      return _data[i - 1];
+    }
+  }
+
+  //
+  // We can change the lock type later. Use a spinlock for now, even if it
+  // is suboptimal in cases where long critical sections have high
+  // contention (IE, lots of tasks trying to insert into the middle of this
+  // List, or any operation that is O(n)).
+  //
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  //
+  // Use a wrapper class to let the List class have a const ref receiver even
+  // when `parSafe` is `true` and the List lock is used.
+  //
+  pragma "no doc"
+  class LockWrapper {
+    
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  /*
+    A List is a lightweight container suitable for building up and iterating
+    over a collection of elements in a structured manner. It is intended to
+    replace the use of arrays to perform "vector-like" operations. Unlike a
```

Comment:
> It is intended to replace the use of arrays to perform "vector-like" operations.

I don't think this should be in the main chpldoc comment. You could put it in a ..note at the end since it's about history.